Src/Test.UnitTests.RE2.Managed/Regex2Tests.cs

```diff
@@ -184,40 +184,87 @@ public void Regex2_MatchAndClear()
-            Assert.True(groupName2Index.ContainsKey("g1"));
-            Assert.Equal(1, groupName2Index["g1"]);
-            Assert.True(groupName2Index.ContainsKey("g2"));
-            Assert.Equal(3, groupName2Index["g2"]);
-            Assert.True(index2GroupName.ContainsKey(1));
-            Assert.Equal("g1", index2GroupName[1]);
-            Assert.True(index2GroupName.ContainsKey(3));
-            Assert.Equal("g2", index2GroupName[3]);
-            Assert.Equal("abc", submatchStrings[0]);
-            Assert.Equal("a", submatchStrings[1]);
-            Assert.Equal("b", submatchStrings[2]);
-            Assert.Equal("c", submatchStrings[3]);
+            List<Dictionary<string, string>> matches;
+
+            bool hasMatches = Regex2.Matches(@"abc", "abc", out matches);
+            Assert.True(hasMatches);
+            Assert.Single(matches);
+            Assert.True(matches[0].ContainsKey("0"));
+            Assert.Equal("abc", matches[0]["0"]);
+        }
+
+        [Fact]
+        public void Regex2_CaptureGroups_NoMatch()
+        {
+            List<Dictionary<string, string>> matches;
+
+            bool hasMatches = Regex2.Matches(@"def", "abc", out matches);
+            Assert.False(hasMatches);
+            Assert.Empty(matches);
+        }
+
+        [Fact]
+        public void Regex2_CaptureGroups_WithGroups()
+        {
+            List<Dictionary<string, string>> matches;
+
+            bool hasMatches = Regex2.Matches(@"(?P<g1>a)(b)(?P<g2>c)", "abc", out matches);
+
+            Assert.True(hasMatches);
+            Assert.Single(matches);
+            Assert.Equal(4, matches[0].Count);
+            Assert.True(matches[0].ContainsKey("0"));
+            Assert.Equal("abc", matches[0]["0"]);
+            Assert.True(matches[0].ContainsKey("g1"));
+            Assert.Equal("a", matches[0]["g1"]);
+            Assert.True(matches[0].ContainsKey("2"));
+            Assert.Equal("b", matches[0]["2"]);
+            Assert.True(matches[0].ContainsKey("g2"));
+            Assert.Equal("c", matches[0]["g2"]);
+        }
+
+        [Fact]
+        public void Regex2_CaptureGroups_VariableLengthGroupNames()
+        {
+            List<Dictionary<string, string>> matches;
+
+            bool hasMatches = Regex2.Matches(@"(?P<a>a)(?P<bb>b)(?P<ccc>c)", "abc", out matches);
+
+            Assert.True(hasMatches);
+            Assert.Single(matches);
+            Assert.Equal(4, matches[0].Count);
+            Assert.True(matches[0].ContainsKey("0"));
+            Assert.Equal("abc", matches[0]["0"]);
+            Assert.True(matches[0].ContainsKey("a"));
+            Assert.Equal("a", matches[0]["a"]);
+            Assert.True(matches[0].ContainsKey("bb"));
+            Assert.Equal("b", matches[0]["bb"]);
+            Assert.True(matches[0].ContainsKey("ccc"));
+            Assert.Equal("c", matches[0]["ccc"]);
+        }
+
+        [Fact]
+        public void Regex2_CaptureGroups_NonOverlapping()
```

Comment:
>Regex2_CaptureGroups_NonOverlapping [](http://example.com/codeflow?start=20&length=35)

can you create a test where we dont find a match?
just to check if the match.count>0 would work fine in this case.

Reply:
Yup, I added the test case `Regex2_CaptureGroups_NoMatch`.